internal/storage/postgres/watering_plan/update.go

```diff
@@ -0,0 +1,51 @@
+package wateringplan
+
+import (
+	"context"
+	"errors"
+
+	"github.com/green-ecolution/green-ecolution-backend/internal/entities"
+	sqlc "github.com/green-ecolution/green-ecolution-backend/internal/storage/postgres/_sqlc"
+	"github.com/green-ecolution/green-ecolution-backend/internal/utils"
+)
+
+func (w *WateringPlanRepository) Update(ctx context.Context, id int32, wpFn ...entities.EntityFunc[entities.WateringPlan]) (*entities.WateringPlan, error) {
+	entity, err := w.GetByID(ctx, id)
+	if err != nil {
+		return nil, w.store.HandleError(err)
+	}
+
+	for _, fn := range wpFn {
+		fn(entity)
+	}
+
+	if err := w.validateWateringPlan(entity); err != nil {
+		return nil, err
+	}
+
+	if err := w.updateEntity(ctx, entity); err != nil {
+		return nil, err
+	}
+
+	return w.GetByID(ctx, entity.ID)
+}
```

Comment:
Maybe we can already use the transaction here (not tested the suggestion)

```suggestion
func (w *WateringPlanRepository) Update(ctx context.Context, id int32, updateFn func(*entities.WateringPlan) (bool, error)) error {
	return w.store.WithTx(ctx, func(s *store.Store) error {
		oldStore := r.store
		defer func() {
			r.store = oldStore
		}()
		r.store = s

		entity, err := w.GetByID(ctx, id)
		if err != nil {
			return err
		}

		if updateFn == nil {
			return errors.New("updateFn is nil")
		}

		updated, err := updateFn(entity)
		if err != nil {
			return err
		}

		if !updated {
			return nil
		}

		return w.updateEntity(ctx, tc)
	})

}
```